feat: Add primitive support for sound api

Timongcraft opened the pull request:
With this PR, the Adventure-provided sound API is **partially** implemented.

It adds the ability to play sounds for a player (from the player itself or other players on the same server) and to stop sounds,
while making the methods fail silently if the sound can't be played/stopped.

What this PR implements:
- A method to play back (custom) sounds for a specific player (from the player itself or other players) for 1.19.3+
(- This also enables the respective methods on the registered server, by forwarding it to the players)
- A method to stop (custom) sounds for players with version 1.19.3+

What this PR **doesn't** implement:
- A method to play back (custom) sounds for a specific player at a position or another entity for players of all versions
- A method to play back (custom) sounds (globally) at a specific location for players of all versions
- A easily way to access a list of vanilla sounds
*- A method to play back (custom) sounds for a specific player for players with version 1.19.2 and below*
*- A method to stop (custom) sounds for players with version 1.19.2 and below*

Clarification:
This PR only implements a sound API for version 1.19.3+ because only in this version is the server able to play a (vanilla) sound for a player without requiring a version-dependent id for the sound.

Timongcraft commented:
TODO:
- [x] Resolve merge conflicts

Timongcraft committed:
feat: Add primitive support for sound api

nicolube commented:
What is the state of this? I would like to have this feature very much.

Timongcraft commented:
From my end, it is finished, although I would think it makes more sense to make the methods fail silently again.

Timongcraft committed:
change to fail silently

fix: implement the correct playSound method
fix: bumped "since" version

Timongcraft commented:
~~We could also look into playing sounds from another player, but if so, I think that should be discussed first.~~

nicolube commented:
@astei Could this be reviewed pls.

nicolube commented:
@4drian3d Since it has been approved, can this also be merged?

TheXplore commented:
What is the state of this? We are currently discussing implementing this ourselves, but we would way prefer having it implemented in upstream.

Timongcraft commented:
Apart from maybe adding Javadoc to the RegisteredServer, **this** PR is ready from my end.

TheXplore commented:
@4drian3d Do you think additional Javadoc entries are required? Is there an estimate when this might get merged?

Timongcraft committed:
chore: update 1.21.5

nicolube commented:
@electronicboy 
Sry for the ping, but what is blocking this from being merged?

kezz commented on proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java:
```diff
@@ -1026,6 +1030,31 @@ void setClientBrand(final @Nullable String clientBrand) {
+  public void playSound(@NotNull Sound sound, @NotNull Sound.Emitter emitter) {
+    Preconditions.checkNotNull(sound, "sound");
+    Preconditions.checkNotNull(emitter, "emitter");
+    Preconditions.checkArgument(emitter.equals(Sound.Emitter.self()), "non-self emitter not supported");
```
Better to fail silently, that's more adventure

kezz replied:
Or possibly just ignore the emitter and play normally

kezz replied:
Also, any reason you only override play sound with an emitter?

Timongcraft replied:
> Also, any reason you only override play sound with an emitter?

Yes, because the other one indicates it uses the position of the player, which wouldn't follow it. So I moved the implemented method [here](https://github.com/PaperMC/Velocity/pull/1422/commits/6ec69dc8b3a6223f30fe32e7fe837de940dd1d0e).
Audience javadoc for normal playSound method: https://javadoc.io/doc/net.kyori/adventure-api/latest/net/kyori/adventure/audience/Audience.html#playSound(net.kyori.adventure.sound.Sound)

Timongcraft replied:
But also if we would add more emitters in the future I thought it would make sense to enforce the use of the self emitter, should we really ignore it?

kezz replied:
Ah yes, and velocity doesn't track the location so we can't use that packet. 

It's API contract for unsupported adventure methods to not throw exceptions.

Timongcraft replied:
Alright, then I'll make it silent was well

Timongcraft replied:
~~Are null checks exempt? Because I see them in some places or should everything fail silently?~~

Timongcraft replied:
@kezz, what do you say?
(also to the other conversation below)

kezz commented on api/src/main/java/com/velocitypowered/api/proxy/Player.java:
```diff
@@ -403,18 +403,26 @@ default void playSound(@NotNull Sound sound, double x, double y, double z) {
-   * <b>This method is not currently implemented in Velocity
-   * and will not perform any actions.</b>
+   * <p>Note: This method is currently only implemented for players from version 1.19.3 and above
+   * and requires a present {@link #getCurrentServer}. Additionally, it only supports {@link Sound.Emitter#self()} for now.
```
If we're tracking all the entity IDs now, other emitters could be supported right? Not a requirement for this PR, just a future thought.

Timongcraft replied:
Yea, like other players but I found out this:
> I did not add support to use other players as emitter, as custom sounds played like that will just appear like if they were coming from the player itself.

And thought it wasn't worth it, right?

kezz replied:
That sounds like you were testing with sounds with the wrong number of channels - they definitely should follow the player.

Timongcraft replied:
wdym? I used the player entity sound packet with a entity id of another player, in I think 1.21.4, and it just acted like the sound was coming from the player I sent the packet to. But I could test again.

Timongcraft replied:
Yea, in my testing (new setup, 1.21.4, with paper and with changed velocity code) when using another player as emitter it only is played from the other player if it is 1. a vanilla sound, and 2. uses the `stream` parameter like `music_disc.cat` but not `minecraft:item.goat_horn.sound.2`.
And thus I did not find it useful and too complicated to understand to include.

kezz replied:
If you have a stereo sound it will never play following the entity and instead will play globally. https://bugs.mojang.com/browse/MC-146721

Timongcraft replied:
Ah thanks, didn't know about that.
I'll test and implement it again.
Would it make sense to add that to the adventure javadoc, I'd open a pr if you want?

kezz replied:
Would be a good addition to the Sound class yes.

Timongcraft replied:
> Would be a good addition to the Sound class yes.

Why the '[Sound](https://github.com/KyoriPowered/adventure/blob/1b8f5e418ebb26b12b6f3755d16bfaa7ada8a548/api/src/main/java/net/kyori/adventure/sound/Sound.java)' class instead of [the specific 'playSound' method in 'Audience'](https://github.com/KyoriPowered/adventure/blob/1b8f5e418ebb26b12b6f3755d16bfaa7ada8a548/api/src/main/java/net/kyori/adventure/audience/Audience.java#L668)?

Timongcraft committed:
chore: enforce adventure's policy of not throwing exceptions on unsupported actions

Timongcraft committed:
feat: allow sounds to be played from other players (on the same server)

Timongcraft commented:
Even though the sound should be ignored for invalid entity ids, I still choose to test for the current server and thus require it on the receiver because afaik they are not that unique and if there's some other entity with that id exists, it could cause unwanted consequences.

Timongcraft committed:
chore(fix): add missing getters/setters in packets

Krymonota commented:
Is there any chance that this could be merged soon? I'd love to use it! :)

Timongcraft commented:
> Is there any chance that this could be merged soon? I'd love to use it! :)

I'd wait for 1.21.6 and see how to handle the implementation of Adventure for [the new "UI" sound category](https://www.minecraft.net/en-us/article/minecraft-1-21-6-pre-release-3).

Edit: Seems like it was just added to the end of the enum, in vanilla and adventure, so it shouldn't be a problem.

Timongcraft committed:
chore: update 1.21.6
chore: added own notes to playSound method, as adventure moved them to the Sound class

Timongcraft committed:
chore: cleanup

Timongcraft committed:
fix: ignore invalid sound source
fix: sound source error on wrong version

TheXplore commented:
Hi, is there a current estimate when this might get merged? :)

Timongcraft committed:
chore: prettify key writing

4drian3d committed:
Implement missing Player#playSound(Sound)

4drian3d reviewed (approved):
lgtm
It works quite well

Timongcraft reviewed (commented):
I did not implement the Player#playSound(Sound) method as adventure defines it like this:
> Plays a sound at the location of the recipient of the sound.

Since we can't fulfill that, would it be fine to break that contract by replacing the javadoc?

https://github.com/KyoriPowered/adventure/blob/-/api/src/main/java/net/kyori/adventure/audience/Audience.java#L629C6-L631

4drian3d commented:
> I did not implement the Player#playSound(Sound) method as adventure defines it like this:
> 
> > Plays a sound at the location of the recipient of the sound.
> 
> Since we can't fulfill that, would it be fine to break that contract by replacing the javadoc?
> 
> https://github.com/KyoriPowered/adventure/blob/-/api/src/main/java/net/kyori/adventure/audience/Audience.java#L629C6-L631

_I'll be honest, I tried this pull request for 30 minutes without knowing why the sound wasn't playing, since stopSound was working for me, until I saw that the `#playSound(Sound)` method wasn't implemented_

From a plugin developer's perspective, this method should be implemented, but from a contract compliance perspective, it's okay not to implement it as required by the contract. With this in mind, good documentation at https://docs.papermc.io/velocity/dev/pitfalls/#audience-operations-are-not-fully-supported would suffice, so I am now reverting my change

4drian3d committed:
Reverted Player#playSound(Sound) implementation

Also, improved documentation related to #playSound mehtods

Timongcraft commented:
What do you say about stuffing override dialog api methods into this pr as well for documentation purposes?

Timongcraft committed:
chore(jd): mark dialog operations unsupported

Timongcraft committed:
chore: update 1.21.9
